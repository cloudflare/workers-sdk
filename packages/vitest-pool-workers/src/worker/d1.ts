import type { D1Migration } from "../shared/d1";

function isD1Database(v: unknown): v is D1Database {
	return (
		typeof v === "object" &&
		v !== null &&
		v.constructor.name === "D1Database" &&
		"prepare" in v &&
		typeof v.prepare === "function" &&
		"batch" in v &&
		typeof v.batch === "function" &&
		"exec" in v &&
		typeof v.exec === "function"
	);
}

function isD1Migration(v: unknown): v is D1Migration {
	return (
		typeof v === "object" &&
		v !== null &&
		"name" in v &&
		typeof v.name === "string" &&
		"queries" in v &&
		Array.isArray(v.queries) &&
		v.queries.every((query) => typeof query === "string")
	);
}
function isD1Migrations(v: unknown): v is D1Migration[] {
	return Array.isArray(v) && v.every(isD1Migration);
}

export async function applyD1Migrations(
	db: D1Database,
	migrations: D1Migration[],
	migrationsTableName = "d1_migrations"
) {
	if (!isD1Database(db)) {
		throw new TypeError(
			"Failed to execute 'applyD1Migrations': parameter 1 is not of type 'D1Database'."
		);
	}
	if (!isD1Migrations(migrations)) {
		throw new TypeError(
			"Failed to execute 'applyD1Migrations': parameter 2 is not of type 'D1Migration[]'."
		);
	}
	// noinspection SuspiciousTypeOfGuard
	if (typeof migrationsTableName !== "string") {
		throw new TypeError(
			"Failed to execute 'applyD1Migrations': parameter 3 is not of type 'string'."
		);
	}

	// Create migrations table if it doesn't exist
	const schema = `CREATE TABLE IF NOT EXISTS ${migrationsTableName} (
		id         INTEGER PRIMARY KEY AUTOINCREMENT,
		name       TEXT UNIQUE,
		applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
	);`;
	await db.prepare(schema).run();

	// Find applied migrations
	const appliedMigrationNamesResult = await db
		.prepare(`SELECT name FROM ${migrationsTableName};`)
		.all<{ name: string }>();
	const appliedMigrationNames = appliedMigrationNamesResult.results.map(
		({ name }) => name
	);

	// Apply un-applied migrations
	const insertMigrationStmt = db.prepare(
		`INSERT INTO ${migrationsTableName} (name) VALUES (?);`
	);
	for (const migration of migrations) {
		if (appliedMigrationNames.includes(migration.name)) {
			continue;
		}

		const queries = migration.queries.map((query) => db.prepare(query));
		queries.push(insertMigrationStmt.bind(migration.name));
		await db.batch(queries);
	}
}
