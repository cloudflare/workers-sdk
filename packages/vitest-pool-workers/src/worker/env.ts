import assert from "node:assert";

export let env: Record<string, unknown>;
export function setEnv(newEnv: Record<string, unknown>) {
	env = newEnv;
}

export function getSerializedOptions(): SerializedOptions {
	assert(typeof __vitest_worker__ === "object", "Expected global Vitest state");
	const options = __vitest_worker__.config?.poolOptions?.miniflare;
	// `options` should always be defined when running tests
	assert(options !== undefined, "Expected serialised options");
	return options;
}
