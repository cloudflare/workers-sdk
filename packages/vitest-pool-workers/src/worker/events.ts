// `workerd` doesn't allow these internal classes to be constructed directly.
// To replicate this behaviour require this unique symbol to be specified as the
// first constructor argument. If this is missing, throw `Illegal invocation`.
const kConstructFlag = Symbol("kConstructFlag");

// =============================================================================
// `ExecutionContext`
// =============================================================================

const kWaitUntil = Symbol("kWaitUntil");
class ExecutionContext {
	// https://github.com/cloudflare/workerd/blob/v1.20231218.0/src/workerd/api/global-scope.h#L168
	[kWaitUntil]: unknown[] = [];

	constructor(flag: typeof kConstructFlag) {
		if (flag !== kConstructFlag) throw new TypeError("Illegal constructor");
	}

	waitUntil(promise: unknown) {
		if (!(this instanceof ExtendableEvent)) {
			throw new TypeError("Illegal invocation");
		}
		this[kWaitUntil].push(promise);
	}

	passThroughOnException(): void {}
}
export function createExecutionContext(): ExecutionContext {
	return new ExecutionContext(kConstructFlag);
}
export async function getWaitUntil<T extends unknown[]>(
	ctx: ExecutionContext
): Promise<T> {
	// noinspection SuspiciousTypeOfGuard
	if (!(ctx instanceof ExecutionContext)) {
		throw new TypeError(
			"Failed to execute 'getWaitUntil': parameter 1 is not of type 'ExecutionContext'.\n" +
				"You must call 'createExecutionContext()' to get an 'ExecutionContext' instance."
		);
	}
	const waitUntil = ctx[kWaitUntil];
	let length = 0;
	let result = [] as unknown as T;
	while (length !== waitUntil.length) {
		length = waitUntil.length;
		result = (await Promise.all(ctx[kWaitUntil])) as T;
	}
	return result;
}

// =============================================================================
// `ScheduledController`
// =============================================================================

const kRetry = Symbol("kRetry");
class ScheduledController {
	// https://github.com/cloudflare/workerd/blob/v1.20231218.0/src/workerd/api/scheduled.h#L35
	readonly scheduledTime!: number;
	readonly cron!: string;
	[kRetry] = true;

	constructor(flag: typeof kConstructFlag, options?: FetcherScheduledOptions) {
		if (flag !== kConstructFlag) throw new TypeError("Illegal constructor");

		const scheduledTime = Number(options?.scheduledTime ?? Date.now());
		const cron = String(options?.cron ?? "");

		// Match `JSG_READONLY_INSTANCE_PROPERTY` behaviour
		Object.defineProperties(this, {
			scheduledTime: {
				get() {
					return scheduledTime;
				},
			},
			cron: {
				get() {
					return cron;
				},
			},
		});
	}

	noRetry(): void {
		if (!(this instanceof ScheduledController)) {
			throw new TypeError("Illegal invocation");
		}
		this[kRetry] = false;
	}
}
export function createScheduledController(
	options?: FetcherScheduledOptions
): ScheduledController {
	if (options !== undefined && typeof options !== "object") {
		throw new TypeError(
			"Failed to execute 'createScheduledController': parameter 1 is not of type 'ScheduledOptions'."
		);
	}
	return new ScheduledController(kConstructFlag, options);
}
export async function getScheduledResult(
	ctrl: ScheduledController,
	ctx: ExecutionContext
): Promise<FetcherScheduledResult> {
	// noinspection SuspiciousTypeOfGuard
	if (!(ctrl instanceof ScheduledController)) {
		throw new TypeError(
			"Failed to execute 'getScheduledResult': parameter 1 is not of type 'ScheduledController'.\n" +
				"You must call 'createScheduledController()' to get a 'ScheduledController' instance."
		);
	}
	// noinspection SuspiciousTypeOfGuard
	if (!(ctx instanceof ExecutionContext)) {
		throw new TypeError(
			"Failed to execute 'getScheduledResult': parameter 2 is not of type 'ExecutionContext'.\n" +
				"You must call 'createExecutionContext()' to get an 'ExecutionContext' instance."
		);
	}
	await getWaitUntil(ctx);
	return { outcome: "ok", noRetry: !ctrl[kRetry] };
}

// =============================================================================
// `MessageBatch`
// =============================================================================

const kAck = Symbol("kAck");
const kRetryAll = Symbol("kRetryAll");
const kAckAll = Symbol("kAckAll");
class QueueMessage /* Message */ {
	// https://github.com/cloudflare/workerd/blob/v1.20231218.0/src/workerd/api/queue.h#L113
	readonly #controller: QueueController;
	readonly id!: string;
	readonly timestamp!: Date;
	readonly body!: unknown;
	[kRetry] = false;
	[kAck] = false;

	constructor(
		flag: typeof kConstructFlag,
		controller: QueueController,
		message: ServiceBindingQueueMessage
	) {
		if (flag !== kConstructFlag) throw new TypeError("Illegal constructor");
		this.#controller = controller;

		const id = String(message.id);

		let timestamp: Date;
		// noinspection SuspiciousTypeOfGuard
		if (typeof message.timestamp === "number") {
			timestamp = new Date(message.timestamp);
		} else if (message.timestamp instanceof Date) {
			timestamp = new Date(message.timestamp.getTime()); // Prevent external mutations
		} else {
			throw new TypeError(
				"Incorrect type for the 'timestamp' field on 'ServiceBindingQueueMessage': the provided value is not of type 'date'."
			);
		}

		if ("serializedBody" in message) {
			throw new TypeError(
				"Cannot use `serializedBody` with `createMessageBatch()`"
			);
		}
		const body = structuredClone(message.body); // Prevent external mutations

		// Match `JSG_READONLY_INSTANCE_PROPERTY` behaviour
		Object.defineProperties(this, {
			id: {
				get() {
					return id;
				},
			},
			timestamp: {
				get() {
					return timestamp;
				},
			},
			body: {
				get() {
					return body;
				},
			},
		});
	}

	retry() {
		if (!(this instanceof QueueMessage)) {
			throw new TypeError("Illegal invocation");
		}
		if (this.#controller[kRetryAll]) return;
		if (this.#controller[kAckAll]) {
			console.warn(
				`Received a call to retry() on message ${this.id} after ackAll() was already called. ` +
					"Calling retry() on a message after calling ackAll() has no effect."
			);
			return;
		}
		if (this[kAck]) {
			console.warn(
				`Received a call to retry() on message ${this.id} after ack() was already called. ` +
					"Calling retry() on a message after calling ack() has no effect."
			);
			return;
		}
		this[kRetry] = true;
	}

	ack() {
		if (!(this instanceof QueueMessage)) {
			throw new TypeError("Illegal invocation");
		}
		if (this.#controller[kAckAll]) return;
		if (this.#controller[kRetryAll]) {
			console.warn(
				`Received a call to ack() on message ${this.id} after retryAll() was already called. ` +
					"Calling ack() on a message after calling retryAll() has no effect."
			);
			return;
		}
		if (this[kRetry]) {
			console.warn(
				`Received a call to ack() on message ${this.id} after retry() was already called. ` +
					"Calling ack() on a message after calling retry() has no effect."
			);
			return;
		}
		this[kAck] = true;
	}
}
class QueueController /* MessageBatch */ {
	// https://github.com/cloudflare/workerd/blob/v1.20231218.0/src/workerd/api/queue.h#L198
	readonly queue!: string;
	readonly messages!: QueueMessage[];
	[kRetryAll] = false;
	[kAckAll] = false;

	constructor(
		flag: typeof kConstructFlag,
		queueOption: string,
		messagesOption: ServiceBindingQueueMessage[]
	) {
		if (flag !== kConstructFlag) throw new TypeError("Illegal constructor");

		const queue = String(queueOption);
		const messages = messagesOption.map(
			(message) => new QueueMessage(kConstructFlag, this, message)
		);

		// Match `JSG_READONLY_INSTANCE_PROPERTY` behaviour
		Object.defineProperties(this, {
			queue: {
				get() {
					return queue;
				},
			},
			messages: {
				get() {
					return messages;
				},
			},
		});
	}

	retryAll() {
		if (!(this instanceof QueueController)) {
			throw new TypeError("Illegal invocation");
		}
		if (this[kAckAll]) {
			console.warn(
				"Received a call to retryAll() after ackAll() was already called. " +
					"Calling retryAll() after calling ackAll() has no effect."
			);
			return;
		}
		this[kRetryAll] = true;
	}

	ackAll() {
		if (!(this instanceof QueueController)) {
			throw new TypeError("Illegal invocation");
		}
		if (this[kRetryAll]) {
			console.warn(
				"Received a call to ackAll() after retryAll() was already called. " +
					"Calling ackAll() after calling retryAll() has no effect."
			);
			return;
		}
		this[kAckAll] = true;
	}
}
export function createMessageBatch(
	queueName: string,
	messages: ServiceBindingQueueMessage[]
): MessageBatch {
	if (arguments.length === 0) {
		// `queueName` will be coerced to a `string`, but it must be defined
		throw new TypeError(
			"TypeError: Failed to execute 'createMessageBatch': parameter 1 is not of type 'string'."
		);
	}
	if (!Array.isArray(messages)) {
		throw new TypeError(
			"TypeError: Failed to execute 'createMessageBatch': parameter 2 is not of type 'Array'."
		);
	}
	return new QueueController(kConstructFlag, queueName, messages);
}
export async function getQueueResult(
	batch: QueueController,
	ctx: ExecutionContext
): Promise<FetcherQueueResult> {
	// noinspection SuspiciousTypeOfGuard
	if (!(batch instanceof QueueController)) {
		throw new TypeError(
			"Failed to execute 'getQueueResult': parameter 1 is not of type 'MessageBatch'.\n" +
				"You must call 'createMessageBatch()' to get a 'MessageBatch' instance."
		);
	}
	// noinspection SuspiciousTypeOfGuard
	if (!(ctx instanceof ExecutionContext)) {
		throw new TypeError(
			"Failed to execute 'getQueueResult': parameter 2 is not of type 'ExecutionContext'.\n" +
				"You must call 'createExecutionContext()' to get an 'ExecutionContext' instance."
		);
	}
	await getWaitUntil(ctx);

	const explicitRetries: string[] = [];
	const explicitAcks: string[] = [];
	for (const message of batch.messages) {
		if (message[kRetry]) explicitRetries.push(message.id);
		if (message[kAck]) explicitAcks.push(message.id);
	}
	return {
		outcome: "ok",
		retryAll: batch[kRetryAll],
		ackAll: batch[kAckAll],
		explicitRetries,
		explicitAcks,
	};
}
