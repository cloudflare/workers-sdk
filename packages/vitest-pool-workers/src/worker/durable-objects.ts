import assert from "node:assert";
import { env as globalEnv, getSerializedOptions } from "./env";
import { importModule, mustGetResolvedMainPath } from "./import";

const CF_KEY_ACTION = "vitestPoolWorkersDurableObjectAction";

let nextActionId = 0;
const kUseResponse = Symbol("kUseResponse");
const actionResults = new Map<number /* id */, unknown>();

function isDurableObjectNamespace(v: unknown): v is DurableObjectNamespace {
	return (
		typeof v === "object" &&
		v !== null &&
		v.constructor.name === "DurableObjectNamespace" &&
		"newUniqueId" in v &&
		typeof v.newUniqueId === "function" &&
		"idFromName" in v &&
		typeof v.idFromName === "function" &&
		"idFromString" in v &&
		typeof v.idFromString === "function" &&
		"get" in v &&
		typeof v.get === "function"
	);
}
function isDurableObjectStub(v: unknown): v is DurableObjectStub {
	return (
		typeof v === "object" &&
		v !== null &&
		(v.constructor.name === "DurableObject" ||
			v.constructor.name === "WorkerRpc") &&
		"fetch" in v &&
		typeof v.fetch === "function" &&
		"id" in v &&
		typeof v.id === "object"
	);
}

// Whilst `sameIsolatedNamespaces` depends on `getSerializedOptions()`,
// `isolateDurableObjectBindings` is derived from the user Durable Object
// config. If this were to change, the Miniflare options would change too
// restarting this worker. This means we only need to compute this once, as it
// will automatically invalidate when needed.
let sameIsolatedNamespaces: DurableObjectNamespace[] | undefined;
function getSameIsolateNamespaces(): DurableObjectNamespace[] {
	if (sameIsolatedNamespaces !== undefined) return sameIsolatedNamespaces;
	const options = getSerializedOptions();
	if (options.isolateDurableObjectBindings === undefined) return [];
	sameIsolatedNamespaces = options.isolateDurableObjectBindings.map((name) => {
		const namespace = globalEnv[name];
		assert(
			isDurableObjectNamespace(namespace),
			`Expected ${name} to be a DurableObjectNamespace binding`
		);
		return namespace as DurableObjectNamespace;
	});
	return sameIsolatedNamespaces;
}

function assertSameIsolate(stub: DurableObjectStub) {
	// Make sure our special `cf` requests get handled correctly and aren't
	// routed to user fetch handlers
	const idString = stub.id.toString();
	const namespaces = getSameIsolateNamespaces();
	// Try to recreate the stub's ID using each same-isolate namespace.
	// `idFromString()` will throw if the ID is not for that namespace.
	// If a call succeeds, we know the ID is for an object in this isolate.
	for (const namespace of namespaces) {
		try {
			namespace.idFromString(idString);
			return;
		} catch {}
	}
	// If no calls succeed, we know the ID is for an object outside this isolate,
	// and we won't be able to use the `actionResults` map to share data.
	throw new Error(
		"Durable Object test helpers can only be used with stubs pointing to objects defined within the same worker."
	);
}

export async function runInDurableObject<O extends DurableObject, R>(
	stub: DurableObjectStub,
	callback: (instance: O, state: DurableObjectState) => R | Promise<R>
): Promise<R> {
	if (!isDurableObjectStub(stub)) {
		throw new TypeError(
			"Failed to execute 'runInDurableObject': parameter 1 is not of type 'DurableObjectStub'."
		);
	}
	if (typeof callback !== "function") {
		throw new TypeError(
			"Failed to execute 'runInDurableObject': parameter 2 is not of type 'function'."
		);
	}

	assertSameIsolate(stub);
	const id = nextActionId++;
	actionResults.set(id, callback);

	const response = await stub.fetch("http://x", {
		cf: { [CF_KEY_ACTION]: id },
	});
	// `result` may be `undefined`
	assert(actionResults.has(id), `Expected action result for ${id}`);
	const result = actionResults.get(id);
	actionResults.delete(id);
	if (result === kUseResponse) {
		return response as R;
	} else if (response.ok) {
		return result as R;
	} else {
		throw result;
	}
}

async function runAlarm(instance: DurableObject, state: DurableObjectState) {
	const alarm = await state.storage.getAlarm();
	if (alarm === null) return false;
	await state.storage.deleteAlarm();
	await instance.alarm?.();
	return true;
}
export function runDurableObjectAlarm(
	stub: DurableObjectStub
): Promise<boolean /* ran */> {
	if (!isDurableObjectStub(stub)) {
		throw new TypeError(
			"Failed to execute 'runDurableObjectAlarm': parameter 1 is not of type 'DurableObjectStub'."
		);
	}
	return runInDurableObject(stub, runAlarm);
}

type DurableObjectConstructor = {
	new (state: DurableObjectState, env: Env): DurableObject;
};
type DurableObjectParameters<K extends keyof DurableObject> = Parameters<
	NonNullable<DurableObject[K]>
>;

// Wrapper for user Durable Object classes defined in this worker,
// intercepts and handles action requests
class DurableObjectWrapper implements DurableObject {
	instanceConstructor?: DurableObjectConstructor;
	instance?: DurableObject;

	constructor(
		readonly state: DurableObjectState,
		readonly env: Env,
		readonly className: string
	) {}

	async ensureInstance(): Promise<DurableObject> {
		const mainPath = mustGetResolvedMainPath("Durable Object");
		// `ensureInstance()` may be called multiple times concurrently.
		// We're assuming `importModule()` will only import the module once.
		const mainModule = await importModule(this.env, mainPath);
		const constructor = mainModule[this.className];
		if (typeof constructor !== "function") {
			throw new Error(
				`${mainPath} does not export a ${this.className} Durable Object`
			);
		}
		this.instanceConstructor ??= constructor as DurableObjectConstructor;
		if (this.instanceConstructor !== constructor) {
			// Unlikely to hit this case if aborting all Durable Objects at the
			// start/end of each test
			await this.state.blockConcurrencyWhile<never>(() => {
				// Throw inside `blockConcurrencyWhile()` to abort this object
				throw new Error(
					`${mainPath} changed, invalidating this Durable Object. ` +
						"Please retry the `DurableObjectStub#fetch()` call."
				);
			});
			assert.fail("Unreachable");
		}
		if (this.instance === undefined) {
			this.instance = new this.instanceConstructor(this.state, this.env);
			// Wait for any `blockConcurrencyWhile()`s in the constructor to complete
			await this.state.blockConcurrencyWhile(async () => {});
		}
		return this.instance;
	}

	async fetch(request: Request): Promise<Response> {
		// Make sure we've initialised user code
		const instance = await this.ensureInstance();

		// If this is an internal Durable Object action, handle it...
		const actionId = request.cf?.[CF_KEY_ACTION];
		if (typeof actionId === "number") {
			try {
				const callback = actionResults.get(actionId);
				assert(
					typeof callback === "function",
					`Expected callback for ${actionId}`
				);
				const result = await callback(instance, this.state);
				// If the callback returns a `Response`, we can't pass it back to the
				// caller through `actionResults`. If we did that, we'd get a `Cannot
				// perform I/O on behalf of a different Durable Object` error if we
				// tried to use it. Instead, we set a flag in `actionResults` that
				// instructs the caller to use the `Response` returned by
				// `DurableObjectStub#fetch()` directly.
				if (result instanceof Response) {
					actionResults.set(actionId, kUseResponse);
					return result;
				} else {
					actionResults.set(actionId, result);
				}
				return new Response(null, { status: 204 });
			} catch (e) {
				actionResults.set(actionId, e);
				return new Response(null, { status: 500 });
			}
		}

		// Otherwise, pass through to the user code
		if (instance.fetch === undefined) {
			throw new Error("Handler does not export a fetch() function.");
		}
		return instance.fetch(request);
	}

	async alarm(...args: DurableObjectParameters<"alarm">) {
		const instance = await this.ensureInstance();
		return instance.alarm?.(...args);
	}
	async webSocketMessage(...args: DurableObjectParameters<"webSocketMessage">) {
		const instance = await this.ensureInstance();
		return instance.webSocketMessage?.(...args);
	}
	async webSocketClose(...args: DurableObjectParameters<"webSocketClose">) {
		const instance = await this.ensureInstance();
		return instance.webSocketClose?.(...args);
	}
	async webSocketError(...args: DurableObjectParameters<"webSocketError">) {
		const instance = await this.ensureInstance();
		return instance.webSocketError?.(...args);
	}
}

export function createDurableObjectWrapper(
	className: string
): DurableObjectConstructor {
	return class extends DurableObjectWrapper {
		constructor(state: DurableObjectState, env: Env) {
			super(state, env, className);
		}
	};
}
