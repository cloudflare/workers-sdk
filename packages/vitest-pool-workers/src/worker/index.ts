import assert from "node:assert";
import events from "node:events";
import * as vm from "node:vm";
import {
	setEnv,
	importModule,
	maybeHandleImportRequest,
	mustGetResolvedMainPath,
} from "cloudflare:test-internal";
import * as devalue from "devalue";
import type { VitestExecutor as VitestExecutorType } from "vitest/execute";

const originalConsole = console;

const originalSetTimeout = globalThis.setTimeout;
const originalClearTimeout = globalThis.clearTimeout;

// HACK: `vitest/dist/vendor/vi.js` attempts to call `setTimeout` when setting
// up global mocks. Unfortunately, the runner Durable Object's IO context isn't
// preserved through `import()` so this fails. To get around this, look for
// the `setTimeout()` call and return a recognisable timeout value that's still
// `number` typed.
// @ts-expect-error __promisify__ types only required for Node.js
globalThis.setTimeout = (...args: Parameters<typeof setTimeout>) => {
	if (
		args[1] === 0 &&
		args[0].toString().replace(/\s/g, "") === "function(){returnundefined;}"
	) {
		return -0.5;
	}
	return originalSetTimeout.apply(globalThis, args);
};
// @ts-expect-error overload types not compatible
globalThis.clearTimeout = (...args: Parameters<typeof clearTimeout>) => {
	if (args[0] === -0.5) return;
	return originalClearTimeout.apply(globalThis, args);
};

// Wraps a `WebSocket` with a Node `MessagePort` like interface
class WebSocketMessagePort extends events.EventEmitter {
	constructor(private readonly socket: WebSocket) {
		super();
		socket.addEventListener("message", this.#onMessage);
		socket.accept();
	}

	#onMessage = (event: MessageEvent) => {
		assert(typeof event.data === "string");
		// originalConsole.log("Worker received message from pool...", event.data);
		const parsed = devalue.parse(event.data);
		this.emit("message", parsed);
	};

	postMessage(data: unknown) {
		// originalConsole.log("Worker sending message to pool...", data);
		const stringified = devalue.stringify(data);
		this.socket.send(stringified);
	}
}

interface JsonError {
	message?: string;
	name?: string;
	stack?: string;
	cause?: JsonError;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function reduceError(e: any): JsonError {
	return {
		name: e?.name,
		message: e?.message ?? String(e),
		stack: e?.stack,
		cause: e?.cause === undefined ? undefined : reduceError(e.cause),
	};
}

export class RunnerObject implements DurableObject {
	executor: VitestExecutorType | undefined;

	constructor(_state: DurableObjectState, env: Env) {
		// @ts-expect-error `_setUnsafeEval()` is an internal method
		vm._setUnsafeEval(env.__VITEST_POOL_WORKERS_UNSAFE_EVAL);

		// Strip internal bindings from user facing `env`
		const userEnv: Record<string, unknown> = { ...env };
		delete userEnv.__VITEST_POOL_WORKERS_RUNNER_OBJECT;
		delete userEnv.__VITEST_POOL_WORKERS_UNSAFE_EVAL;
		setEnv(userEnv);
	}

	async handleVitestRunRequest(request: Request): Promise<Response> {
		assert.strictEqual(request.headers.get("Upgrade"), "websocket");
		const { 0: poolSocket, 1: poolResponseSocket } = new WebSocketPair();

		const workerDataHeader = request.headers.get("MF-Vitest-Worker-Data");
		assert(workerDataHeader !== null);
		const wd: unknown = devalue.parse(workerDataHeader);
		assert(typeof wd === "object" && wd !== null);
		assert("filePath" in wd && typeof wd.filePath === "string");
		assert("name" in wd && typeof wd.name === "string");
		assert("data" in wd && typeof wd.data === "object" && wd.data !== null);

		const port = new WebSocketMessagePort(poolSocket);
		try {
			const module = await import(wd.filePath);

			// HACK: Internally, the Vitest worker calls `startViteNode()`, which
			// constructs a singleton `VitestExecutor`. `VitestExecutor` is a subclass
			// of `ViteNodeRunner`, which is how the worker communicates with the
			// Vite server. We'd like access to this singleton so we can transform and
			// import code with Vite ourselves (e.g. for user worker's default exports
			// and Durable Objects). Unfortunately, Vitest doesn't publicly export the
			// `startViteNode()` function. Instead, we monkeypatch a `VitestExecutor`
			// method we know is called to get the singleton. :see_no_evil:
			// TODO(soon): see if we can get `startViteNode()` (https://github.com/vitest-dev/vitest/blob/8d183da4f7cc2986d11c802d16bacd221fb69b96/packages/vitest/src/runtime/execute.ts#L45)
			//  exported in `vitest/execute` (https://github.com/vitest-dev/vitest/blob/main/packages/vitest/src/public/execute.ts)
			const { VitestExecutor } = await import("vitest/execute");
			const originalResolveUrl = VitestExecutor.prototype.resolveUrl;
			// eslint-disable-next-line @typescript-eslint/no-this-alias
			const that = this;
			VitestExecutor.prototype.resolveUrl = function (...args) {
				that.executor = this;
				return originalResolveUrl.apply(this, args);
			};

			(wd.data as { port: WebSocketMessagePort }).port = port;
			module[wd.name](wd.data)
				.then(() => {
					poolSocket.close(1000, "Done");
				})
				.catch((e: unknown) => {
					const error = reduceError(e);
					originalConsole.error("Error running worker:", error.stack);
					poolSocket.close(1011, "Internal Error");
				});
		} catch (e) {
			const error = reduceError(e);
			originalConsole.error("Error initialising worker:", error.stack);
			return Response.json(error, {
				status: 500,
				headers: { "MF-Experimental-Error-Stack": "true" },
			});
		}

		return new Response(null, { status: 101, webSocket: poolResponseSocket });
	}

	async fetch(request: Request): Promise<Response> {
		// This will fail if this is an import request, and we haven't called
		// `handleVitestRunRequest()` yet
		const response = await maybeHandleImportRequest(this.executor, request);
		if (response !== undefined) return response;

		return this.handleVitestRunRequest(request);
	}
}

function createHandlerWrapper<K extends keyof ExportedHandler<Env>>(
	key: K
): NonNullable<ExportedHandler<Env>[K]> {
	return async (thing: unknown, env: Env, ctx: ExecutionContext) => {
		const mainPath = mustGetResolvedMainPath("service");
		const mainModule = await importModule(env, mainPath);
		const defaultExport =
			typeof mainModule === "object" &&
			mainModule !== null &&
			"default" in mainModule &&
			mainModule.default;
		const handlerFunction =
			typeof defaultExport === "object" &&
			defaultExport !== null &&
			key in defaultExport &&
			(defaultExport as Record<string, unknown>)[key];
		if (typeof handlerFunction === "function") {
			return handlerFunction.call(defaultExport, thing, env, ctx);
		} else {
			let message = `Handler does not export a ${key}() function.`;
			if (!defaultExport) {
				message +=
					"\nIt looks like your main module is missing a `default` export. `@cloudflare/vitest-pool-workers` does not support service workers." +
					"\nPlease migrate to the modules format: https://developers.cloudflare.com/workers/reference/migrate-to-module-workers.";
			}
			throw new Error(message);
		}
	};
}
const handler: Required<ExportedHandler<Env>> = {
	fetch: createHandlerWrapper("fetch"),
	tail: createHandlerWrapper("tail"),
	trace: createHandlerWrapper("trace"),
	scheduled: createHandlerWrapper("scheduled"),
	test: createHandlerWrapper("test"),
	email: createHandlerWrapper("email"),
	queue: createHandlerWrapper("queue"),
};
export default handler;

// Re-export user Durable Object wrappers
export * from "__VITEST_POOL_WORKERS_USER_OBJECT";
