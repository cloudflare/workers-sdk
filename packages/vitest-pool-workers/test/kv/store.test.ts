import {
	env,
	fetchMock,
	runInDurableObject,
	runDurableObjectAlarm,
	createExecutionContext,
	getWaitUntil,
} from "cloudflare:test";
import { describe, it, expect, afterAll, beforeAll } from "vitest";
import worker, { transformResponse, Counter } from "./worker";

beforeAll(() => {
	fetchMock.activate();
	fetchMock.disableNetConnect();
	fetchMock
		.get("https://example.com")
		.intercept({ path: "/" })
		.reply(200, "data");
});
afterAll(() => fetchMock.assertNoPendingInterceptors());

describe("kv", () => {
	console.log("Defining tests..."); // Check log out of test has correct stack

	it("user agents", () => {
		console.log({ env });
		expect(navigator.userAgent).toBe("Cloudflare-Workers");
	});

	it("stores in KV", async () => {
		await env.TEST_NAMESPACE.put("key", "value");
		expect(await env.TEST_NAMESPACE.get("key")).toBe("value");
	});

	it("uses seeded data", async () => {
		expect(await env.TEST_NAMESPACE.get("seeded")).toBe("ðŸŒ±");
	});

	it("stores in Durable Objects", async () => {
		const id = env.COUNTER.idFromName("a");
		const stub = env.COUNTER.get(id);
		let response = await stub.fetch("http://x/abc");
		expect(await response.json()).toMatchObject({ value: 1 });
		response = await stub.fetch("http://x/abc");
		expect(await response.json()).toMatchObject({ value: 2 });
	});
	it("stores in Durable Object instances", async () => {
		const id = env.COUNTER.idFromName("a");
		const stub = env.COUNTER.get(id);

		let response = await stub.fetch("http://x");
		expect(await response.json()).toMatchObject({ value: 1 }); // Isolation!
		response = await runInDurableObject(stub, (instance: Counter) => {
			expect(instance).toBeInstanceOf(Counter);
			return instance.fetch(new Request("http://x"));
		});
		expect(await response.json()).toMatchObject({ value: 2 });

		const value = await runInDurableObject(stub, (_instance, state) =>
			state.storage.get<number>("/")
		);
		expect(value).toBe(2);

		response = await runInDurableObject(stub, (instance) =>
			instance.fetch(new Request("http://x"))
		);
		expect(await response.json()).toMatchObject({ value: 3 });
	});

	it("calls Durable Object alarm", async () => {
		const id = env.COUNTER.idFromName("alarm-test");
		const stub = env.COUNTER.get(id);
		await runInDurableObject(stub, async (_instance, state) => {
			await state.storage.setAlarm(Date.now() + 10_000);
		});
		expect(await runDurableObjectAlarm(stub)).toBe(true);
		const alarmResult = await runInDurableObject(
			stub,
			(instance: Counter) => instance.alarmPromise
		);
		expect(alarmResult).toBe(42);
		expect(await runDurableObjectAlarm(stub)).toBe(false);
	});

	it("uses other object", async () => {
		const id = env.OTHER.idFromName("other-test");
		const stub = env.OTHER.get(id);
		const response = await stub.fetch("http://x");
		expect(await response.text()).toBe("other Durable Object body");
		// Can only use Durable Object helpers for same-isolate objects
		await expect(runInDurableObject(stub, () => {})).rejects.toThrow();
	});

	it("fetches", async () => {
		const request = new Request<unknown, IncomingRequestCfProperties>(
			"http://localhost"
		);
		const ctx = createExecutionContext();
		const response = await worker.fetch(request, env, ctx);
		await getWaitUntil(ctx);
		expect(await response.text()).toBe("body:http://localhost");
	});

	it("transforms", async () => {
		const response = transformResponse(
			new Response('<a href="http://example.com"></a>')
		);
		expect(await response.text()).toBe('<a href="https://example.com"></a>');
	});

	it("sends request to self", async () => {
		const res = await env.SELF.fetch("http://localhost");
		expect(await res.text()).toMatchInlineSnapshot(`"body:http://localhost"`);
	});

	it("mocks fetch requests", async () => {
		const res = await fetch("https://example.com");
		expect(await res.text()).toMatchSnapshot("mocked-data");
	});
});
