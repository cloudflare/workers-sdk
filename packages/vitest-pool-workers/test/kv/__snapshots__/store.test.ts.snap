// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`kv > mocks fetch requests > mocked-data 1`] = `"data"`;
